File: node/src/admin/realms.ts
import type KcAdminClient from '@keycloak/keycloak-admin-client'
import type RealmRepresentation from '@keycloak/keycloak-admin-client/lib/defs/realmRepresentation.js'
import { call, requireFound } from './call.js'

/**
 * 렐름 생성/조회/삭제 파사드. `RealmsResource`는 realm 스코프가 아니라 렐름 이름을 인자로
 * 받는다(생성 payload의 `realm` 필드가 대상 렐름 이름). 공식 admin-client의 `realms`
 * 리소스를 감싸며 {@link call}로 경계 변환한다.
 */
export class RealmsResource {
  /**
   * @internal `AdminClient`가 조립한다 — 소비자 생성 경로가 아니다. `@internal` + `stripInternal`로
   * 방출 `.d.ts`에서 지워, 하위 `KcAdminClient` 타입이 공개 표면에 오르지 않게 한다(§4).
   */
  constructor(private readonly kc: KcAdminClient) {}

  /** 새 렐름을 생성한다. `representation.realm`이 새 렐름의 이름이다. */
  async create(representation: RealmRepresentation): Promise<void> {
    await call(() => this.kc.realms.create(representation))
  }

  /** 렐름 이름으로 조회한다. 없으면 `KeycloakNotFoundError`. */
  async get(realmName: string): Promise<RealmRepresentation> {
    return requireFound(
      await call(() => this.kc.realms.findOne({ realm: realmName })),
      `Realm not found: ${realmName}`,
    )
  }

  /** 호출자가 볼 수 있는 렐름 전부. 서비스 계정은 보통 자기 렐름만 본다 — 전체를 가정하지 말 것. */
  async list(): Promise<RealmRepresentation[]> {
    return call(() => this.kc.realms.find())
  }

  /**
   * 현재 이름으로 주소를 잡아 렐름을 갱신한다. `representation.realm`에 새 이름을 주면 rename이다.
   *
   * ⚠️ 경로(`realmName`)와 body(`representation`)를 **합치지 말 것** — 경로 인자를 body의 이름으로
   * 덮어쓰면 rename이 조용한 no-op이 된다.
   */
  async update(realmName: string, representation: RealmRepresentation): Promise<void> {
    await call(() => this.kc.realms.update({ realm: realmName }, representation))
  }

  async delete(realmName: string): Promise<void> {
    await call(() => this.kc.realms.del({ realm: realmName }))
  }
}

File: kotlin/src/main/kotlin/io/github/xzawed/keycloak/admin/Realms.kt
package io.github.xzawed.keycloak.admin

import org.keycloak.representations.idm.RealmRepresentation

// Realms.kt — RealmsResource: 공식 admin-client의 org.keycloak.admin.client.resource.RealmsResource를
// 감싸며 모든 호출을 adminCall(AdminClient.kt)로 경계 변환한다(Java RealmsResource 동형).
//
// ⚠️ API 편차(Java와 동일): RealmsResource에는 get(String)/delete(String)가 직접 존재하지 않는다
// (create(RealmRepresentation)만 있고 반환 타입도 void). 개별 렐름 조회/삭제는 realm(name)으로 얻은
// RealmResource의 toRepresentation()/remove()로 위임한다. get()은 대상이 없으면 삼키지 않고
// KeycloakAdminException.NotFound를 그대로 전파한다(UsersResource.get()과 동일한 정책).
public class RealmsResource internal constructor(
    private val delegate: org.keycloak.admin.client.resource.RealmsResource,
) {
    public suspend fun create(representation: RealmRepresentation) {
        adminCall { delegate.create(representation) }
    }

    public suspend fun get(realmName: String): RealmRepresentation = adminCall { delegate.realm(realmName).toRepresentation() }

    /** 호출한 서비스 계정이 볼 수 있는 렐름만 돌아온다(보통 자기 렐름). */
    public suspend fun list(): List<RealmRepresentation> = adminCall { delegate.findAll() }

    /**
     * 현재 이름으로 주소를 잡아 렐름을 갱신한다. representation.realm에 새 이름을 주면 rename이다.
     *
     * ⚠️ 경로(realmName)와 body(representation)를 합치지 말 것 — 합치면 rename이 조용한 no-op이 된다.
     */
    public suspend fun update(
        realmName: String,
        representation: RealmRepresentation,
    ) {
        adminCall { delegate.realm(realmName).update(representation) }
    }

    public suspend fun delete(realmName: String) {
        adminCall { delegate.realm(realmName).remove() }
    }
}
